planning/motion_velocity_planner/autoware_motion_velocity_planner_common/src/polygon_utils.cpp

```diff
@@ -175,33 +200,53 @@ std::vector<geometry_msgs::msg::Pose> calculate_error_poses(
 
 std::optional<std::pair<geometry_msgs::msg::Point, double>> get_collision_point(
   const std::vector<TrajectoryPoint> & traj_points, const std::vector<Polygon2d> & traj_polygons,
-  const geometry_msgs::msg::Point obj_position, const rclcpp::Time obj_stamp,
+  const geometry_msgs::msg::Point obj_position, [[maybe_unused]] const rclcpp::Time obj_stamp,
   const Polygon2d & obj_polygon, const double x_offset_to_bumper)
 {
-  const auto collision_info =
-    get_collision_index(traj_points, traj_polygons, obj_position, obj_stamp, obj_polygon);
-  if (!collision_info) {
-    return std::nullopt;
-  }
+  assert(traj_points.size() == traj_polygons.size());
+  const double obj_maximum_length = boost::geometry::perimeter(obj_polygon) * 0.5;
 
-  const auto bumper_pose = autoware_utils_geometry::calc_offset_pose(
-    traj_points.at(collision_info->first).pose, x_offset_to_bumper, 0.0, 0.0);
+  std::optional<std::pair<geometry_msgs::msg::Point, double>> nearest_collision{std::nullopt};
 
-  std::optional<double> max_collision_length = std::nullopt;
-  std::optional<geometry_msgs::msg::Point> max_collision_point = std::nullopt;
-  for (const auto & poly_vertex : collision_info->second) {
-    const double dist_from_bumper =
-      std::abs(autoware_utils_geometry::inverse_transform_point(poly_vertex.point, bumper_pose).x);
+  for (size_t i = 0; i < traj_polygons.size(); ++i) {
+    const double ego_maximum_length = boost::geometry::perimeter(traj_polygons.at(i)) * 0.5;
+    const double center_dist =
+      autoware_utils_geometry::calc_distance2d(traj_points.at(i).pose.position, obj_position);
+    if (center_dist > obj_maximum_length + ego_maximum_length) {
+      continue;
+    }
+
+    const double current_arc_length = motion_utils::calcSignedArcLength(traj_points, 0, i);
+
+    if (
+      nearest_collision.has_value() &&
+      std::abs(current_arc_length - x_offset_to_bumper) > std::abs(nearest_collision->second)) {
+      break;
+    }
 
-    if (!max_collision_length.has_value() || dist_from_bumper > *max_collision_length) {
-      max_collision_length = dist_from_bumper;
-      max_collision_point = poly_vertex.point;
+    std::vector<Polygon2d> collision_polygons;
+    boost::geometry::intersection(traj_polygons.at(i), obj_polygon, collision_polygons);
+    if (collision_polygons.empty()) {
+      continue;
+    }
+    const auto bumper_pose = autoware_utils_geometry::calc_offset_pose(
+      traj_points.at(i).pose, x_offset_to_bumper, 0.0, 0.0);
+    const auto max_penetration = find_max_penetration_point(
+      collision_polygons, bumper_pose, traj_points.at(i).pose.position.z);
```

Comment:
If you accept proposed change above :+1: 
```suggestion
    const auto max_penetration = find_furthest_collision_point(
      collision_polygons, bumper_pose, traj_points.at(i).pose.position.z);
```